### ISSUE TITLE
Use the same extensions commit for both build and *.openjdk tests in PR builds

### ISSUE THREAD

**Issue by jdmpapin** [eclipse-openj9/openj9#20121]
It's currently possible for a PR build to build with one extensions repo commit and then test using a different one. This can cause spurious test failures, e.g. https://github.com/eclipse-openj9/openj9/pull/20091#issuecomment-2332156034, and there might be other ways in which the openjdk tests expect to be kept consistent with the build

I don't know whether this can also happen in non-PR build+test jobs, e.g. nightlies

**Comment by keithc-ca** [comment]
It's not clear that this is an easily achievable goal: github doesn't even guarantee that pull request merge commits _exist_, but so far we've been lucky. The bigger problem, though, is that those merge commits are a _moving target_: refs/pull/20091/merge, for example, represents the result of merging the head of the target branch (`master` in the example) with the branch behind the pull request (`vTableSlot-cpIndex` in the example). If the target branch advances, likely due to merging of some unrelated pull request, then that merge commit changes as well.

Unless we want to build machinery that explicitly creates named merge commits (and cease using the implicit merge commits) then the possibility of running tests that change because of merging some other pull request will continue to exist. If we use new branches for building and testing, it's possible that builds and testing succeed (in that named branch) but will fail once the pull request is merged. We will also have the problem of managing all those extra branches.

I think the problem we have now (which seems to occur rather infrequently) is preferable to the problems that come with the alternative.

**Comment by jdmpapin** [comment]
I'm not talking about completely avoiding racing against other merges. That would require something like [bors](https://github.com/graydon/bors), which I'm not suggesting we use (and which seems to have been superseded, interesting). Given how often bugs are intermittent, tests passing now is no guarantee that they won't fail later on, even with the exact same code. Anyway, short of aggressively serializing like bors does, we're stuck with this possibility anyway:
> it's possible that builds and testing succeed (in that named branch) but will fail once the pull request is merged.

We could have two incompatible PRs, run PR builds that are both clean, and then merge both, and run into the problem only after the second merge. This isn't related to named branches. The possibility arises simply because each PR can be tested independently of the other without ever testing them in tandem before both are merged.

Rather, this issue is just about making a PR build use a single consistent revision of a given extensions repo for both the build and test steps. Really it should only use a single revision of every repo it uses, including openj9 and omr, but hopefully those are already OK. I don't know of any reason we should expect that to be difficult. The PR build already determines which commits to use for building, and it passes them to the build jobs, which build those commits specifically. The only difference I'm asking for is that if any of the tests are taken from the same repos, they should be taken from the same commits that were used in the build step. There is very little benefit to picking up changes that happen to have been merged between build and test. And while yes it's rare to run into a problem due to this, the problem I hit was very puzzling, and it was a huge pain to debug. 

Maybe my point is more obvious if we consider allowing a different part of the source to vary. Imagine that in a PR build, when it came time to build the JIT, we grabbed the latest source (potentially merged with the PR) for the *JIT only*. This could get us a slightly more up-to-date JIT source tree, but whenever it did there would be a risk that it's incompatible with e.g. the older VM source. Hopefully we can agree that that would be a bad idea. But when running sanity.openjdk in a PR build, that's effectively the situation with the extensions repo right now. We use part of it from one commit to build with, and then we use the tests from a different commit

**Comment by keithc-ca** [comment]
> making a PR build use a single consistent revision of a given extensions repo for both the build and test steps

Please explain how you propose to accomplish that. The SHA that was used at build time may not exist (because refs/pull/20091/merge has moved).

**Comment by jdmpapin** [comment]
Oh, does GitHub GC those commits that quickly? I'll just assume that they do

At least for openj9 PR builds, most of the time there's no "depends" clause for the extensions repo, and the extensions commit is in the openj9 branch already, where it will stay

However, on closer inspection it seems I was wrong about this:
> The PR build already determines which commits to use for building, and it passes them to the build jobs, which build those commits specifically.

The parameters look like they're doing that, but apparently `OPENJ9_SHA` is actually the merge parent from master, and the build job freshly fetches `refs/pull/<n>/merge`

It would take some more doing than I thought, but I think it should still be possible to use consistent revisions for the whole PR build, even in cases where we're incorporating changes from a PR. We'd need to generate a file that we can then make available to the child jobs. I figure that must be possible, since a build job generates archives that are available later to the corresponding test job(s). Instead of using `pull/<n>/merge` directly, we could use it just to find its parent commits. The first is the tip of the target branch (or at least, it was recently), and the second is the tip of the PR branch (similar). We know the target branch commit isn't going anywhere, and we can use it together with the PR branch to create a patch file, e.g.
```
$ git fetch origin pull/20093/merge:pr-merge
$ git format-patch --stdout pr-merge^1..pr-merge^2 >pr.patch
```
When starting the child jobs we'd specify the tip of the target branch as a SHA, i.e. `git rev-parse pr-merge^1` as observed in the top-level PR build job, and a URL for the patch (maybe zipped). The child jobs would check out the commit specified by the SHA (which is still there) and then apply the patch with `git am` to recreate the commits from the PR. They'd technically be new commits with different hashes, but at the end of the series we should get exactly the same contents as the merge commit had. We could even apply the patches in the top-level job and compare just to effectively assert that the contents will turn out identical

**Comment by keithc-ca** [comment]
Yes, GitHub generates the commit behind `ref/pull/<n>/merge` automatically (but not officially guaranteed), but it isn't GC that's the concern. GitHub (force-)updates the branch if the target branch advances, but it seems it doesn't always do it automatically - sometimes you have to view the pull request in your browser to prompt GitHub to update the related merge commit.

I am strongly opposed to anything I can imagine involving "patch" files. Builds should consume commits directly from the relevant repositories.

The content of (each of) the branches involved (including `ref/pull/<n>/merge` could be pushed to branches that would be static for the duration of the build and test jobs, but that involves creating (much) more complicated build processes. I don't see that as a good use of resources (developers time).

**Comment by jdmpapin** [comment]
> The content of (each of) the branches involved (including `ref/pull/<n>/merge` could be pushed to branches that would be static for the duration of the build and test jobs, but that involves creating (much) more complicated build processes. I don't see that as a good use of resources (developers time).

Yes, I agree that that approach would be unnecessarily complex, and I haven't suggested it, nor would I

> I am strongly opposed to anything I can imagine involving "patch" files. Builds should consume commits directly from the relevant repositories.

`format-patch` and `am` are built into git, but OK. Come to think of it, I think we could do pretty much the same thing without patch files, and in the process get rid of the weirdness with `pull/<n>/merge` getting out of date

The idea this time is to keep using a merge commit in much the same way as now, but to stop using `pull/<n>/merge` entirely, and instead have each job do the merge itself. The top-level job determines the tips of the target and PR branches. For the PR branch we could use `pull/<n>/head` if that's reliably up to date, or we could just use the downstream fork as a remote directly. It passes both hashes down to the child jobs. Each child job checks out the target branch commit based on the SHA it was given, merges the PR branch into it based on the other SHA, and proceeds with the resulting merge commit. The top-level job could trial the merge beforehand to make sure that it will succeed

The target branch commit isn't going anywhere. The only problem could be if the PR branch is updated in the middle of the PR build. If the PR commit disappears, we could just fail. I think for anyone encountering such a failure it would be clear that it was caused by the original commit going away. But GitHub does seem to keep old commits from PR branches that have been superseded by a force-push, at least those that it links from the PR page

**Comment by keithc-ca** [comment]
A single build combines content from at least three repositories:
* omr
* openj9
* an extension repository

and perhaps others:
* aqa-systemtest
* aqa-tests
* TKG
* etc. ...

Unless explicit, **stable** names (i.e. branches or tags) are pushed to _each_ repository (or a fork) involved this problem will not go away. Today we rely on `pull/<n>/merge` for the pull request itself, and (typically) the default branches for each of the other repositories; none of those names is **stable** (each can move if the related branch moves).

To my mind, this problem occurs infrequently enough that the simple solution (i.e. try again) is sufficient.
